subsys/net/lib/sockets/sockets_tls.c

```diff
@@ -879,7 +879,7 @@ static int tls_mbedtls_handshake(struct tls_context *context, bool block)
 		}
 
 		NET_ERR("TLS handshake error: -%x", -ret);
-		ret = -ECONNABORTED;
+		ret = -ECONNREFUSED;
```

Comment:
>  Assuming this translates to the actual socket errno

It does

>  there's no listening socket at the corresponding destination address/port. Existing ECONNABORTED seems like much better match for "TLS handshake error".

This errno (from this line) is also used when the client tries to initiate a dtls session and no server is running at all. At least in that case ECONNREFUSED seems to be better that ECONNABORTED.

Reply:
> This errno (from this line) is also used when the client tries to initiate a dtls session and no server is running at all. At least in that case ECONNREFUSED seems to be better than ECONNABORTED.

Then it would be better to do more detailed translation of mbedTLS -> POSIX errno error codes, because otherwise changes like that improve one case but may lead to more confusion with other cases in the future (for example, if there's cipher/key mismatches and connection can't proceed due to that, then returning ECONNREFUSED for it is in no way better, because ECONNREFUSED means "no server running at all").

At the very least, that should go into a separate commit with the above explanation why it was done, so someone later could improve it (instead of risking that it may be just flip-flopped on seeing a case like described above).

Reply:
>>     because ECONNREFUSED means "no server running at all"

> Where does that come from? errno.h states:

That comes from deeply ingrained experience with standard sockets.

> #define ECONNREFUSED 111 /**< Connection refused */
> And a more detailed description from connect() man pages:
>       ECONNREFUSED
>              The target address was not listening for connections or refused the connection request.

To me, that translates to: "server was listening, but had its listen() backlog full".

> Yes, the handshake may fail for multiple reasons, but in my opinon "Connection refused" is a good summary that the connection did not succeed.

Again, with my socket experience, I'd find such error confusing (literally, would likely be misguided by it, thinking that there's a problem with server just accepting connections, not with TLS handshake). As I mentioned, one can only so much shovel TLS features into BSD Sockets semantics, so there always will be compromises. I tried to argue a point that this one isn't the best. I didn't try to -1 though, because in the end, only you, as the TLS sockets maintainer, can decide what makes sense with the overall picture in mind.